### PR TITLE
Add movements history grouping in product page

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | In Stock tab of the expermiental Product page v2, include stock movements from orders as grouped rows in history list. 
| Type?             | new feature
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | No issue.
| How to test?      | Alternate types of stock movements for a given product:<ol><li>FO: create a few orders for the given product</li><li>BO: change new orders status to "Shipped" or "Delivered"</li><li>BO > Product page v2 > Stock: edit stock quantity of the given product to trace stock movements</li><li>Repeat steps 1-3 a few times in order to alternate stock movement types</li></ol>Then go to BO > New Product page > Stock tab to take a look at the last stock movements history:<ul><li>Stock movements from orders should be grouped together into single rows</li><li>Each manual stock edition should be displayed as single rows, no grouping</li></ul>Apply the same procedure to test a given combination.<br/>Please note you can also create new orders from BO, and immediatly set the appropriate order status before saving (steps 1 & 2).<br /><br />The feature also takes into account the returned products, the grouped row between each edition is the sum of all stock movements related to orders wether they are positive or negative. In any case the number of displayed rows is limited to 5<br/><br/>**Note**: this PR must be tested in **single shop only** you shouldn't even enable the multishop mode, this feature will be developed later in another PR
| Possible impacts? | Once this PR is validated and merged, we should consider this update in the autoupgrade module https://github.com/PrestaShop/autoupgrade/pull/494 to keep the DB structure consistent

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27407)

Capture d’écran 2022-06-03 à 12 07 06

#### Breaking Change

- `CombinationStockUpdater` has two new parameters in its constructor
- `CombinationFormDataProvider` has two new parameters in its constructor

### CODE REVIEW

File: src/Adapter/Product/Combination/Update/CombinationStockUpdater.php
Comment (PululuK) [inline src/Adapter/Product/Combination/Update/CombinationStockUpdater.php:79]: Use configuration interface
`use PrestaShop\PrestaShop\Core\Domain\Configuration\ShopConfigurationInterface;`

File: src/Adapter/Product/Combination/Update/CombinationStockUpdater.php
Comment (PululuK) [inline src/Adapter/Product/Combination/Update/CombinationStockUpdater.php:175]:
```suggestion

            // Update reserved and physical quantity for this stock
            $this->stockAvailableRepository->updatePhysicalProductQuantity(
                new StockId((int) $stockAvailable->id),
                new OrderStateId($this->configuration->getInt('PS_OS_ERROR', null, ShopConstraint::shop((int) $stockAvailable->id_shop))),
                new OrderStateId($this->configuration->getInt('PS_OS_CANCELED', null, ShopConstraint::shop((int) $stockAvailable->id_shop)))
            );
```

File: src/Adapter/Product/Combination/Update/CombinationStockUpdater.php
Comment (jolelievre) [reply src/Adapter/Product/Combination/Update/CombinationStockUpdater.php:79]: I can't because I need to use the ShopContraint parameter which is not in the interface yet.

File: src/Adapter/Product/Combination/Update/CombinationStockUpdater.php
Comment (jolelievre) [reply src/Adapter/Product/Combination/Update/CombinationStockUpdater.php:175]: I can't do this all the getters from the ParameterBag don't implement the third ShopConstraint parameter As long as we don't update the interface with this parameter we are kind of stuck and need to use the Adapter service

Besides `getInt` `getBool` and so on are not part of the `ConfigurationInterface` either so your two comments here are a bit contradictory We will have to one day update this interface and ideally force the shop constraint parameter which would remove any ambiguity related to the multishop feature but it's a big breaking change which we can't afford yet

File: src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php
Comment (PululuK) [inline src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php:313]: Put this code in a `AbstractStockAvailableRepository` to avoid code duplicate with `StockAvailableRepository::updateReservedProductQuantity`

Comment (jolelievre) [reply src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php:313]: I wish I could but it's not that easy because `StockAvailableRepository` and `StockAvailableMultiShopRepository` don't have the same parent (`AbstractObjectModelRepository` and `AbstractMultiShopObjectModelRepository` for the other) so I can't create a common abstract class.

The reason why we have those two different repositories is that the multishop development is not finished everywhere on the product page, the strategy was to handle it feature by feature. For stock management it is done for regular product but not for combination for example that's why we introduce some temporary multishop repositories which are indeed doublons with the standard initial ones.

We need both right now because they don't have the same interfaces (multishop repositories force the use of a ShopConstraint parameter), the idea is to one day only have one of the two when everything is done and merge them together but sadly we didn't have time to finish the multishop dev everywhere so we have to keep those two different repositories for now.

Comment (jolelievre) [reply src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php:313]: Since the multishop will likely be finished in v8.1 it also means that we will have to keep both services to avoid a breaking change, so the two repositories will be merged into one only in v9 now I guess.

That's why for now we have indeed duplicate code between the multishop repositories and the regular ones (the same problem applies in other domains with multishop repositories). Except by using traits I don't see how we can fix this, but many maintainers are not comfortable with use of traits and they would also be trait bound to be removed at some point which will also imply a breaking change. So no perfect solution here.

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Use an SCSS variable please :)

File: src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php
Comment (matks) [reply src/Adapter/Product/Stock/Repository/StockAvailableMultiShopRepository.php:313]: Just for the record I ❤️ traits 😛 

But it seems it's not everybody 😄

File: src/Adapter/Product/Stock/Repository/StockMovementRepository.php
Comment (matks) [inline src/Adapter/Product/Stock/Repository/StockMovementRepository.php:109]: 😅 wouah

File: src/Core/Domain/Product/Stock/Query/AbstractGetStockMovements.php
Comment (matks) [inline src/Core/Domain/Product/Stock/Query/AbstractGetStockMovements.php]: Nice-to-have: I'm not 100% convinced of the need of an abstract for 3 setters

**Review by matks — DISMISSED** [review]
OK for PHP part. This PR is a reminder of how much warehouse business rules are complex

**Review by matthieu-rolland — COMMENTED** [review]
@jolelievre would you have a link toward the specs of this feature ?

It looks good code-wise, but I clearly don't understand everything about the logic in this PR, which keeps me from approving this PR...

File: src/Core/Domain/Product/Stock/Query/AbstractGetStockMovements.php
Comment (jolelievre) [reply src/Core/Domain/Product/Stock/Query/AbstractGetStockMovements.php]: Agreed, I'm gonna remove the abstract

**Comment by jolelievre** [comment]
> @jolelievre would you have a link toward the specs of this feature ?
> 
> It looks good code-wise, but I clearly don't understand everything about the logic in this PR, which keeps me from approving this PR...

@matthieu-rolland the description was there to try and explain the feature as well as possible, along with the screenshot But you're right I'm gonna ask @MatShir where the specs are?

And just as I write this I get an answer 😅 Here are the specs which have not been merged yet: https://github.com/PrestaShop/prestashop-specs/pull/302

**Comment by jolelievre** [comment]
@matks @matthieu-rolland you should have all the elements for a final approval now ^^

**Review by matks — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]
impressive work 👍

**Comment by jolelievre** [comment]
> impressive work 👍

Most of the work was done by @saulaski, a great thanks to him

**Review by khouloudbelguith — COMMENTED** [review]
Hi @jolelievre,

I found some issues:

**First issue**
1. Go to BO > Advanced Parameters > Team page
2. Edit the employee > try to change the lastname & firstname with a long expressions for example
lastname = `belguithccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccdddddddddddddddddddddddddddddddddddddddddddddcccccccccccc`
firstname = `khououdcccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd`
2. Go to BO > Catalog > Product V2 > Edit a product
3. In the stock tab > edit quantity and save
4. See error 
image
5. Edit a product with combinations > Edit a combinations
6. Set quantity and save
7. See error
image


**Second issue**

1. Go to BO > Catalog > Product V2 > Edit
2. In the stock tab > edit quantity and save
3. Check the employee => OK
4. Edit the employee > try to update the lastname & firstname with valid expressions
5. Go back to BO > Catalog > Products > Edit the last product 
6. In the stock tab 
7. See error => The employee is not updated 

image

If I check Orders view page > Status block: the employee well updated
image

I attached a screen record

https://user-images.githubusercontent.com/16067358/174125821-f7ee3d58-c508-4d0a-9ede-3568c0e8fcd0.mp4

**Third issue**

1. Install this PR with English language & France country
2. Go to BO > Localization > Languages page > check the `Date format (full)` 
3. image
4. Go to BO > Catalog > Product V2 > Edit
5. In the stock tab > edit quantity and save
6. See error > `Date & Time` format is NOK `2022-06-16 18:55:16`

image

If I check Orders view page > Status block: the Date format is ok
image

**Fourth issue**

**_Case1_**
1. Create a product & set the quantity 
image
2. Create an order with this product & change the status Shipped 
image
3. Create a partial refund of this order 
image
4. EDit the product
5. In the Stock tab > See error => The **Shipped products** line is removed & the **Returned products** line is not displayed
image

If I click to view all stock movements => OK 
image

**_Case2_**

1. Create a product with a quantity = 10
2. Create an order with quantity = 2 & update the status to shipped
3. Edit the product & change the quantity to 3
4. Create a partial refund with quantity = 1
If I edit the product > I have this in the Stock tab
image


5. At this step, create an order with this product & quantity = 1 & change order status to shipped
6. Edit the product > In the stock tab > See error => The **Returned products** line is removed 

image

I attached a screen record

https://user-images.githubusercontent.com/16067358/174131281-96095bd5-9345-4ea6-a094-599d05abc26c.mp4

If I check view all stock movements 
image

Thanks!

**Comment by jolelievre** [comment]
Hi @khouloudbelguith 

thanks for the feedback, here are my answers to each of them.

#### First issue
The bug happens because the employee's first name and last name are saved in the row of `stock_mvt` except the DB structures were not in sync the movement table only accepted 32 characters. So I increased this limit and now the bug is gone. I anticipated your feedback about the design now that the long name can be displayed and limited the display toa avoid breaking the UX:
Capture d’écran 2022-06-20 à 15 54 03

I simply prevented it from being too large and if it exceeds the maximum width some ellipsis are used at the end, this is a good enough solution for this PR since this feature has not been designed, if improvement is required it will be in another PR.

I also updated the autoupgrade module so that it updates the DB structure https://github.com/PrestaShop/autoupgrade/pull/494

#### Second issue
For this one as explained above, the stock movements save the employee's name which means it is saved and persisted at the moment the stock movement is performed and we display the data from the `stock_mvt` table not the Employee's current data. Unlike order page which doesn't save the employee's data only the ID.
So I won't change this behaviour as this difference has not been specified and is not blocking, any improvement/modification about this will need to be specified otherwise and handled in another PR.

#### Third issue
This is not related to this PR, we also display dates in other parts of the page and the format is always hardcoded so far. I didn't know about this feature but I'm not even sure it applies to the product page. If the language format needs to be applied it should be on the whole page and it will need specifications as well. So it won't be fixed in this PR and we keep the hardcoded format for now to be consistent with the rest of the page.

#### Fourth issue
The stock changes between employee editions is the sum of all order-related movements which include product bought AND product returned, if the sum is zero there is nothing to show so we don't display anything. This was decided with @MatShir and @TristanLDD so it is currently the expected behaviour.

**Review by matthieu-rolland — DISMISSED** [review]

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (Amoifr) [inline src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php]: could you phpdoc the array type ?

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (Amoifr) [inline src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php:93]: what a trans can do about firstname and lastname ?

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (Amoifr) [inline src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php:108]: It's sad to continue to implements code with arrays and offset names to poor of meaning. May be usign DTO that maps this name with full completely understandable names can help to cut the link with legacy one day.

File: src/Adapter/Product/Stock/QueryHandler/GetCombinationStockMovementsHandler.php
Comment (Amoifr) [inline src/Adapter/Product/Stock/QueryHandler/GetCombinationStockMovementsHandler.php:42]: what kind of array return is expected ?

File: src/Adapter/Product/Stock/QueryHandler/GetProductStockMovementsHandler.php
Comment (Amoifr) [inline src/Adapter/Product/Stock/QueryHandler/GetProductStockMovementsHandler.php:42]: please define returned array

File: tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php
Comment (Amoifr) [inline tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php:269]: why casting value instead defining return type for the method ?

**Review by matks — DISMISSED** [review]
@jolelievre I approve the code in it current state and let you see if you want to also add Amoifr suggestions 👍

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (jolelievre) [reply src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php:93]: In some cultures the lastname and firstname are inversed, so the most adapted feature to handle this is to leave it the the translator. This way in any language you can simply translate `'%firstname% %lastname%'` with `'%lastname% %firstname%'`

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (jolelievre) [reply src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php:108]: Well it's actually what we are doing here, we keep using the array return from the repository which returns the DB result as a flattened array and then we create DTO objects that are used in the query result

I'm not sure having an intermediate DTO structure between the DB result and the query DTO would bring much value in this case.

File: src/Adapter/Product/Stock/QueryHandler/GetCombinationStockMovementsHandler.php
Comment (jolelievre) [reply src/Adapter/Product/Stock/QueryHandler/GetCombinationStockMovementsHandler.php:42]: It returns a `StockMovementEvent[]` as explained in the interface but I forgot to add the `@inheritdoc` here, thanks

File: tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php:269]: Return type is already defined as `Product` which is a legacy ObjectModel in which I'm not 100% confident with the correct typing of each field

**Review by matks — DISMISSED** [review]
OK for me

Do we need to modify something to match the new column size? Like SQL script or structure?

**Comment by jolelievre** [comment]
@matks indeed we do, the PR is already ready https://github.com/PrestaShop/autoupgrade/pull/494

No need to change the db structure file, because stock movement has a Doctrine entity so its DB structure is generated by doctrine, only the upgrade scripts are required.

**Comment by jolelievre** [comment]
@matthieu-rolland can you please re-approve this PR? I just improved the PHPDoc as the feedback suggested

**Review by matks — DISMISSED** [review]

**Review by zuk3975 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
@jolelievre Little thing to fix

```
Error: Method Tests\Unit\Core\Form\IdentifiableObject\DataProvider\ProductFormDataProviderTest::createResultBasedOnQuery() should return 

array<PrestaShop\PrestaShop\Core\Domain\Product\Customization\QueryResult\CustomizationField|PrestaShop\PrestaShop\Core\Domain\Product\FeatureValue\QueryResult\ProductFeatureValue|PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\RelatedProduct|PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\StockMovementEvent>|PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\ProductForEditing|PrestaShop\PrestaShop\Core\Domain\Product\Supplier\QueryResult\ProductSupplierOptions

but returns array<PrestaShop\PrestaShop\Core\Domain\Product\Pack\QueryResult\PackedProductDetails>.
 ------ ----------------------------------------------------------------------- 
  Line   tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataP  
         roviderTest.php                                                        
 ------ -----------------------------------
 ```

**Review by matks — DISMISSED** [review]

**Review by zuk3975 — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hello @jolelievre,

Could you please check the Conflicting files?

Thanks!

**Review by matthieu-rolland — COMMENTED** [review]
400 modified files ? my browser just froze :D

EDIT: ah yes, rebase needed :+1:

File: src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php
Comment (zuk3975) [inline src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementsHandler.php:67]:
```suggestion
    protected function getStockMovementEvents(StockId $stockId, int $offset, int $limit): array
```

File: src/Core/Domain/Product/Stock/QueryHandler/GetProductStockMovementsHandlerInterface.php
Comment (zuk3975) [inline src/Core/Domain/Product/Stock/QueryHandler/GetProductStockMovementsHandlerInterface.php:35]: I wonder why did you rename the QueryResult (StockMovement -> StockMovementEvent), but not the queries and their handlers? Because the same naming reasons applies both for the query and for the result, so if you had a good reason to rename the query result dto, same should apply for the query and the handler.

File: src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php
Comment (zuk3975) [inline src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php]:
```suggestion
            function (StockMovementEvent $stockMovementEvent): array {
```
or
```suggestion
            function (StockMovementEvent $event): array {
```

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (zuk3975) [reply admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: :up:

File: src/Core/Domain/Product/Stock/QueryHandler/GetProductStockMovementsHandlerInterface.php
Comment (jolelievre) [reply src/Core/Domain/Product/Stock/QueryHandler/GetProductStockMovementsHandlerInterface.php:35]: You're right it's not logical, and actually I don't like this Event suffix much So I renamed everything to a simpler `StockMovement` 😅

**Comment by jolelievre** [comment]
@zuk3975 ready for review again ^^

**Review by zuk3975 — DISMISSED** [review]

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]:
```suggestion
    color: $white;
```

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Pretty sure you can find a SCSS var to replace `219, 225, 227`

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: And now, add this to an SCSS variable if no one exist, or use an existing one

**Review by NeOMakinG — DISMISSED** [review]

**Review by zuk3975 — DISMISSED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hello @jolelievre,

**Issue1**

1. Go to BO > Advanced Parameters > New & Experimental Features page & Enable New product page - Single store
2. Create a vector product with downloadable file & quantity = 10 for example
3. Create an order with this product (No shipping is selected)
4. Go to BO > Orders > View the last order created & change the status to "Payment accepted"

The customer, in this case, can download his attachment > I think, in this specific case, we should add a line for the stock movement

5. Change the Order status to shipped 
6. Go to BO > Catalog > Product list V2 > Edit the product created
7. Check the Virtual tab > The expression "Shipped" is not correct with a virtual product

image

Ping @MatShir what do you think?

**Issue2**

_**Case1**_
1. Create two standard products with quantity = 10 "test" & "Copy test" in my case
2. Create a product with those products with quantity = 5, the quantity of each product into the pack = 2
3. Check "Decrement product in pack only"
4. Save 
image
5. Create an order with the pack product (quantity ordered = 1)
6. Change Status to Shipped
7. Go to BO > Catalog > Edit pack product 
8. See error > A new line is added & the quantity is increased NOK 
image
9. Go to BO > Catalog > Edit the products "test" & "Copy test" in my case
10. In the stock tab > See error No line was added about shipped product 
image


_**Case2**_
1. Create two standard products with quantity = 10 "test" & "Copy test" in my case
2. Create a product with those products with quantity = 5, the quantity of each product into the pack = 2
3. Check "Decrement both."
4. Save 
5. Create an order with the pack product (quantity ordered = 1)
6. Change Status to Shipped
9. Go to BO > Catalog > Edit the products "test" & "Copy test" in my case
10. In the stock tab > See error No line was added about shipped product 

Thanks!

**Comment by MatShir** [comment]
**Issue 1**
It will be fixed when → https://github.com/PrestaShop/PrestaShop/issues/27993 will be fixed. 
**Issue 2**
Thanks a lot for the detection issues about the stock movement @khouloudbelguith. I will create a dedicated EPIC about that. It seems that the pack is not working at all :/
But none of the issues are blocking the PR, so if it is @khouloudbelguith okay for you we can validate the PR and create the EPIC? WDYT?

**Review by zuk3975 — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]
front LGTM 🍬

**Review by khouloudbelguith — APPROVED** [review]
Hi @jolelievre,

It is ok :heavy_check_mark: 

- I checked with all types of products
- I tried to create an order from BO / FO 
- I tried to update an order from BO (adding / deleting products with/without cart rules)
- I tried to import products => OK
- I tried to create product via API => OK

```
I found a new issue but as discussed in slack, it is not related to the PR

1. Create a product Gift with a quantity 100
2. Create a cart rule + Restrictions "Product selection = customizable mug" + free gift (the product created previously)
3. Create an order with a gift product (quantity ordered = 1)
4. Go to BO > Orders > Edit the last order created
5. Add the product "customizable mug" (the cart rule is applied automatically and the quantity of gift ordered is updated to 2)
6. Change order status to 2
7. Go to Catalog > Products > check Stock tab of the gift > incorrect shipped is -3, it should be -2
8. Click on view Mouvements
9. The is an incorrect type (Product return)
```
Thanks!

**Pull request merged**
